Confirmed duplicate message from the client if a Throwable is thrown

Example
```
@Route("")
@Theme(Lumo.class)
public class DemoView extends VerticalLayout {

    public DemoView() {
        getElement().addEventListener("click", e -> {
            throw new Error("Fail");
        });
}
```

Clicking on the button logs two exceptions
```
[WARNING] Error for /
java.lang.Error: Fail
	at DemoView.lambda$0(DemoView.java:14)
	at com.vaadin.flow.internal.nodefeature.ElementListenerMap.lambda$fireEvent$0(ElementListenerMap.java:171)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.vaadin.flow.internal.nodefeature.ElementListenerMap.fireEvent(ElementListenerMap.java:171)
	at com.vaadin.flow.server.communication.rpc.EventRpcHandler.handleNode(EventRpcHandler.java:58)
	at com.vaadin.flow.server.communication.rpc.AbstractRpcInvocationHandler.handle(AbstractRpcInvocationHandler.java:60)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocationData(ServerRpcHandler.java:377)
	at com.vaadin.flow.server.communication.ServerRpcHandler.lambda$handleInvocations$0(ServerRpcHandler.java:367)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:367)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:309)
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:89)
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1541)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:345)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
```

which is expected but then the client side gets a 500 back. The client side then tries to re-send the message as it does not know if the message ever reached the server. This will cause another exception to be logged
```
[qtp1532675992-33] ERROR com.vaadin.flow.server.DefaultErrorHandler - 
java.lang.UnsupportedOperationException: Confirmed duplicate message from the client. Expected sync id: 1, got 0. Message start: {"csrfToken":"ab0f5d03-ba71-408a-8e60-9bfc09b8df65","rpc":[{"type":"event","node":3,"event":"click"}],"syncId":0,"clientId":0}
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:302)
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:89)
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1541)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:345)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:835)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1685)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:225)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
```

In FW8, this case is handled as
```
            ui.getSession().getCommunicationManager().repaintAll(ui);

            if (rpcRequest.getClientToServerId() < expectedId) {
                // Just a duplicate message due to a bad connection or similar
                // It has already been handled by the server so it is safe to
                // ignore
                getLogger()
                        .fine("Ignoring old message from the client. Expected: "
                                + expectedId + ", got: "
                                + rpcRequest.getClientToServerId());
```
which seems more appropriate.

In real world, this happened during development when part of the code was not redeployed and caused a `java.lang.NoSuchMethodError` to be thrown

I'm seeing this on Vaadin 14.0.4. I'm using SpringBoot/tomcat as part of my framework. 

As advised by the doco I turned off SpringBoot csrf handling.

Even though I'm getting this error the customer is seeing a error dialog pop up with an internal error indicated. The application however is fine!

I am seeing his exception partially reproduced here:
[ERROR] com.vaadin.flow.server.DefaultErrorHandler - 
java.lang.UnsupportedOperationException: Confirmed duplicate message from the client. Expected sync id: 4, got 3. Message start: {"csrfToken":"0f9df571-94e1-4cf9-ba41-753b5c112731","rpc":[{"type":"event","node":4,"event":"click","data":{"event.shiftKey":false,"event.metaKey":false,"event.detail":1,"event.ctrlKey":false,"event.clientX":295,"event.clientY":881,"event.altKey":false,"event.button":0,"event.screenY":1009,"event.screenX":295}}],"syncId":3,"clientId":3}
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:294)
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:87)
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1540)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:246)
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:95)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:712)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:459)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:352)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:312)
	at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:52)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1039)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:908)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:660)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
...
etc etc

--- a/flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/communication/ServerRpcHandler.java
@@ -279,7 +279,6 @@ public class ServerRpcHandler implements Serializable {
             // it would only get an empty response (because the dirty flags have
             // been cleared on the server) and would be out of sync
 
-            String message;
             if (requestId == expectedId - 1 && Arrays.equals(messageHash,
                     ui.getInternals().getLastProcessedMessageHash())) {
                 /*
@@ -287,28 +286,31 @@ public class ServerRpcHandler implements Serializable {
                  * situation is most likely triggered by a timeout or such
                  * causing a message to be resent.
                  */
-                message = "Confirmed duplicate message from the client.";
+                getLogger().info(
+                        "Ignoring old duplicate message from the client. Expected: "
+                                + expectedId + ", got: " + requestId);
             } else {
-                message = "Unexpected message id from the client.";
+                /*
+                 * If the reason for ending up here is intermittent, then we
+                 * should just issue a full resync since we cannot know the
+                 * state of the client engine.
+                 *
+                 * There are reasons to believe that there are deterministic
+                 * issues that trigger this condition, and we'd like to collect
+                 * more data to uncover anything such before actually
+                 * implementing the resync that would thus hide most symptoms of
+                 * the actual root cause bugs.
+                 */
+                String messageStart = changeMessage;
+                if (messageStart.length() > 1000) {
+                    messageStart = messageStart.substring(0, 1000);
+                }
+                throw new UnsupportedOperationException(
+                        "Unexpected message id from the client."
+                                + " Expected sync id: " + expectedId + ", got "
+                                + requestId + ". Message start: "
+                                + messageStart);
             }
-
-            /*
-             * If the reason for ending up here is intermittent, then we should
-             * just issue a full resync since we cannot know the state of the
-             * client engine.
-             *
-             * There are reasons to believe that there are deterministic issues
-             * that trigger this condition, and we'd like to collect more data
-             * to uncover anything such before actually implementing the resync
-             * that would thus hide most symptoms of the actual root cause bugs.
-             */
-            String messageStart = changeMessage;
-            if (messageStart.length() > 1000) {
-                messageStart = messageStart.substring(0, 1000);
-            }
-            throw new UnsupportedOperationException(
-                    message + " Expected sync id: " + expectedId + ", got "
-                            + requestId + ". Message start: " + messageStart);
         } else {
             // Message id ok, process RPCs
             ui.getInternals().setLastProcessedClientToServerId(expectedId,

--- a/flow-server/src/test/java/com/vaadin/flow/server/communication/ServerRpcHandlerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/communication/ServerRpcHandlerTest.java
@@ -1,6 +1,7 @@
 package com.vaadin.flow.server.communication;
 
 import java.io.IOException;
+import java.io.Reader;
 import java.io.StringReader;
 
 import org.junit.Assert;
@@ -13,10 +14,13 @@ import org.mockito.Mockito;
 import com.vaadin.flow.component.UI;
 import com.vaadin.flow.component.internal.UIInternals;
 import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.internal.MessageDigestUtil;
 import com.vaadin.flow.internal.StateTree;
 import com.vaadin.flow.server.VaadinRequest;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.flow.server.communication.ServerRpcHandler.InvalidUIDLSecurityKeyException;
+import com.vaadin.flow.shared.ApplicationConstants;
 
 public class ServerRpcHandlerTest {
     private VaadinRequest request;
@@ -67,12 +71,50 @@ public class ServerRpcHandlerTest {
                 + "\", \"rpc\":[], \"resynchronize\": true, \"clientId\":1}");
         uiTree.collectChanges(c -> { // clean tree
         });
-        thrown.expect(ServerRpcHandler.ResynchronizationRequiredException.class);
+        thrown.expect(
+                ServerRpcHandler.ResynchronizationRequiredException.class);
 
         // when
         serverRpcHandler.handleRpc(ui, reader, request);
 
         // then
         Assert.assertTrue(uiTree.hasDirtyNodes());
+    }
+
+    @Test
+    public void handleRpc_duplicateMessage_doNotThrhow()
+            throws InvalidUIDLSecurityKeyException, IOException {
+        String msg = "{\"" + ApplicationConstants.CLIENT_TO_SERVER_ID + "\":1}";
+        ServerRpcHandler handler = new ServerRpcHandler() {
+            @Override
+            protected String getMessage(Reader reader) throws IOException {
+                return msg;
+            };
+        };
+
+        ui = new UI();
+        ui.getInternals().setSession(session);
+        ui.getInternals().setLastProcessedClientToServerId(1,
+                MessageDigestUtil.sha256(msg));
+
+        // This invocation shouldn't throw. No other checks
+        handler.handleRpc(ui, Mockito.mock(Reader.class), request);
+    }
+
+    @Test(expected = UnsupportedOperationException.class)
+    public void handleRpc_unexpectedMessage_thrhow()
+            throws InvalidUIDLSecurityKeyException, IOException {
+        ServerRpcHandler handler = new ServerRpcHandler() {
+            @Override
+            protected String getMessage(Reader reader) throws IOException {
+                return "{\"" + ApplicationConstants.CLIENT_TO_SERVER_ID
+                        + "\":1}";
+            };
+        };
+
+        ui = new UI();
+        ui.getInternals().setSession(session);
+
+        handler.handleRpc(ui, Mockito.mock(Reader.class), request);
     }
 }

